kratos/utilities/sensitivity_builder.cpp

```diff
@@ -0,0 +1,627 @@
+ *
+ * Any data associated with the function (e.g., functors or capture by value)
+ * is copied to each thread.
+ */
+template <typename TContainer, typename TCallable>
+void ParallelForEach(TContainer& rContainer, TCallable&& Fun)
+{
+#pragma omp parallel
+    {
+        auto local_fun{Fun};
+#pragma omp for
+        for (int i = 0; i < static_cast<int>(rContainer.size()); ++i)
+        {
+            std::forward<TCallable>(local_fun)(*(rContainer.begin() + i));
+        }
+    }
+}
+
+/**
+ * @brief Contains the sensitivity design and output variables
+ *
+ * The design variable is passed to CalculateSensitivityMatrix()
+ * and CalculatePartialSensitivity() when calculating the local
+ * sensitivity contributions. The local sensitivities are assembled
+ * to the output variable.
+ *
+ * Example 1:
+ * SensitivityVariables<double> vars("THICKNESS");
+ * vars.pDesignVariable->Name(); // "THICKNESS"
+ * vars.pOutputVariable->Name(); // "THICKNESS_SENSITIVITY"
+ *
+ * Example 2:
+ * SensitivityVariables<double> vars("THICKNESS_SENSITIVITY");
+ * vars.pDesignVariable->Name(); // "THICKNESS_SENSITIVITY"
+ * vars.pOutputVariable->Name(); // "THICKNESS_SENSITIVITY"
+ */
+template <class TDataType>
+struct SensitivityVariables
+{
+    const Variable<TDataType>* pDesignVariable = nullptr;
+    const Variable<TDataType>* pOutputVariable = nullptr;
+
+    explicit SensitivityVariables(const std::string& rName)
+    {
+        KRATOS_TRY;
+        const std::string output_suffix = "_SENSITIVITY";
+        pDesignVariable = &KratosComponents<Variable<TDataType>>::Get(rName);
+        if (rName.size() > output_suffix.size() &&
+            std::equal(output_suffix.rbegin(), output_suffix.rend(), rName.rbegin()))
+        {
+            pOutputVariable = pDesignVariable;
+        }
+        else
+        {
+            pOutputVariable =
+                &KratosComponents<Variable<TDataType>>::Get(rName + output_suffix);
+        }
+        KRATOS_CATCH("");
+    }
+};
+
+template <typename TDataType, typename TContainer>
+void AssembleNodalSolutionStepContainerContributions(const SensitivityVariables<TDataType>& rVariables,
+                                                     TContainer& rContainer,
+                                                     AdjointResponseFunction& rResponseFunction,
+                                                     const ProcessInfo& rProcessInfo,
+                                                     double ScalingFactor)
+{
+    KRATOS_TRY;
+    LocalSensitivityBuilder builder;
+    ParallelForEach(rContainer, [&, builder](typename TContainer::data_type& rElement) mutable {
+        auto& r_geom = rElement.GetGeometry();
+        if (RequiresUpdateSensitivities(r_geom) == false)
+            return;
+        builder.CalculateLocalSensitivity(*rVariables.pDesignVariable, rElement,
+                                          rResponseFunction, rProcessInfo);
+        for (auto& r_value : builder.LocalSensitivity)
```

Comment:
Is there no function for multiplying a vector with a scalar?

Reply:
I didn't think about that but there is :+1: I switched to `*=`.